File: app/typescript-ms/financial-accounting-service/src/application/ports/AccountRepository.ts
// src/application/ports/AccountRepository.ts

import { Account } from '../../domain/models/account/Account'
import { AccountCode } from '../../domain/models/account/AccountCode'

/**
 * 勘定科目リポジトリ（Output Port）
 * ドメイン層がデータアクセスに依存しないためのインターフェース
 */
export interface AccountRepository {
  /**
   * 勘定科目を保存する
   * @param account 保存する勘定科目
   * @returns 保存された勘定科目
   */
  save(account: Account): Promise<Account>

  /**
   * すべての勘定科目を取得する
   * @returns 勘定科目一覧（表示順でソート）
   */
  findAll(): Promise<Account[]>

  /**
   * 科目コードで勘定科目を検索する
   * @param accountCode 科目コード
   * @returns 勘定科目、見つからない場合はnull
   */
  findByCode(accountCode: AccountCode): Promise<Account | null>

  /**
   * BSPL区分で勘定科目を取得する
   * @param bsplDistinction BSPL区分（'B' または 'P'）
   * @returns 勘定科目一覧（表示順でソート）
   */
  findByBsplDistinction(bsplDistinction: string): Promise<Account[]>

  /**
   * 取引要素区分で勘定科目を取得する
   * @param transactionDistinction 取引要素区分
   * @returns 勘定科目一覧（表示順でソート）
   */
  findByTransactionDistinction(transactionDistinction: string): Promise<Account[]>

  /**
   * 勘定科目を削除する
   * @param accountCode 削除する勘定科目の科目コード
   */
  deleteByCode(accountCode: AccountCode): Promise<void>

  /**
   * すべての勘定科目を削除する（テスト用）
   */
  deleteAll(): Promise<void>
}

File: app/ruby/accounting_system/app/repositories/account_repository.rb
# frozen_string_literal: true

# インフラストラクチャ層：Output Adapter
# ActiveRecord を使用してデータアクセスを実装
class AccountRepository
  include AccountRepositoryInterface

  # すべての勘定科目を取得
  # @return [Array<Account>]
  def find_all
    Account.order(:display_order, :code).all.to_a
  end

  # 科目コードで勘定科目を検索
  # @param account_code [String] 科目コード
  # @return [Account, nil]
  def find_by_code(account_code)
    Account.find_by(code: account_code)
  end

  # BSPL 区分で勘定科目を検索
  # @param bspl_type [String] BSPL 区分（'B' または 'P'）
  # @return [Array<Account>]
  def find_by_bspl_type(bspl_type)
    Account.where(bspl_type: bspl_type)
           .order(:display_order, :code)
           .to_a
  end

  # 勘定科目を保存（作成または更新）
  # @param account [Account] 勘定科目
  # @return [Account]
  def save(account)
    account.save!
    account
  end

  # 勘定科目を削除
  # @param account_code [String] 科目コード
  # @return [void]
  def delete_by_code(account_code)
    account = find_by_code(account_code)
    account&.destroy!
  end
end
